pkg/parse/root.go

```diff
@@ -261,6 +262,71 @@ func setSourceStatusFields(source *v1beta1.SourceStatus, newStatus *SourceStatus
 	source.LastUpdate = newStatus.LastUpdate
 }
 
+func (p *root) setSourceAnnotations(ctx context.Context, commit string) error {
+	rs := &v1beta1.RootSync{}
+	if err := p.Client.Get(ctx, rootsync.ObjectKey(p.SyncName), rs); err != nil {
+		return status.APIServerError(err, "failed to get RootSync for parser")
+	}
+
+	var patchOperations []map[string]interface{}
+
+	annotations := rs.GetAnnotations()
+	if annotations == nil {
+		patchOperations = append(patchOperations, map[string]interface{}{
+			"op":    "add",
+			"path":  "/metadata/annotations",
+			"value": map[string]string{},
+		})
+		annotations = make(map[string]string)
+	}
+
+	if _, exists := annotations[metadata.SourceCommitAnnotationKey]; exists {
+		patchOperations = append(patchOperations, map[string]interface{}{
+			"op":    "replace",
+			"path":  fmt.Sprintf("/metadata/annotations/%s", escapeJSONPointer(metadata.SourceCommitAnnotationKey)),
+			"value": commit,
+		})
+	} else {
+		patchOperations = append(patchOperations, map[string]interface{}{
+			"op":    "add",
+			"path":  fmt.Sprintf("/metadata/annotations/%s", escapeJSONPointer(metadata.SourceCommitAnnotationKey)),
+			"value": commit,
+		})
+	}
+
+	// Update the source-url annotation based on the source type
+	var newSourceURL string
+	if p.Options.SourceType == configsync.OciSource || p.Options.SourceType == configsync.HelmSource {
+		newSourceURL = p.Options.SourceRepo
+	}
+
+	currentImageURL := rs.GetAnnotations()[metadata.SourceURLAnnotationKey]
+	if newSourceURL != currentImageURL {
+		if newSourceURL == "" {
+			patchOperations = append(patchOperations, map[string]interface{}{
+				"op":   "remove",
+				"path": fmt.Sprintf("/metadata/annotations/%s", metadata.SourceURLAnnotationKey),
+			})
+		} else {
+			patchOperations = append(patchOperations, map[string]interface{}{
+				"op":    "replace",
+				"path":  fmt.Sprintf("/metadata/annotations/%s", metadata.SourceURLAnnotationKey),
+				"value": newSourceURL,
+			})
+		}
+	}
+
+	if len(patchOperations) == 0 {
+		return nil
+	}
+	patchData, err := json.Marshal(patchOperations)
+	if err != nil {
+		return fmt.Errorf("failed to marshal patch operations: %v", err)
+	}
+
+	return p.Client.Patch(ctx, rs, client.RawPatch(types.JSONPatchType, patchData), client.FieldOwner(configsync.FieldManager))
```

Comment:
It is simpler to use `types.MergePatchType`.

```
patch := `{"metadata":{"annotations":{"image-to-sync":"url:digest"}}}`
cl.Patch(ctx, rs,
  client.RawPatch(types.MergePatchType, []byte(patch)),
  client.FieldOwner(configsync.FieldManager)
```

Reply:
Done